### PR TITLE
Add lint for `unused_result_ok`

### PR DESCRIPTION
**Pull request by ithinuel**
This PR adds a lint to capture the use of `expr.ok();` when the result is not _really_ used.

This could be interpreted as the result being checked (like it is with `unwrap()` or `expect`) but
it actually only ignores the result.

`let _ = expr;` expresses that intent better.

This was also mentionned in #8994 (although not being the main topic of that issue).

changelog: [`misleading_use_of_ok`]: Add new lint to capture `.ok();` when the result is not _really_ used.

### CODE REVIEW

File: clippy_lints/src/misleading_use_of_ok.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    #[clippy::version = "1.70.0"]
+    pub MISLEADING_USE_OF_OK,
+    style,
+    "Use of `.ok()` to silence `Result`'s `#[must_use]` is misleading. Use `let _ =` instead."
+}
+declare_lint_pass!(MisleadingUseOfOk => [MISLEADING_USE_OF_OK]);
+
+impl LateLintPass<'_> for MisleadingUseOfOk {
+    fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &Stmt<'_>) {
+        if let StmtKind::Semi(expr) = stmt.kind &&
```
[the remark this answers is not in the document]
Comment (ithinuel) [reply clippy_lints/src/misleading_use_of_ok.rs:37]: Do you have any guidance/example on how to achieve that?

File: clippy_lints/src/misleading_use_of_ok.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            let ExprKind::MethodCall(ok_path, recv, [], ..) = expr.kind //check is expr.ok() has type Result<T,E>.ok(, _)
+            && ok_path.ident.as_str() == "ok"
+            && is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Result)
+        {
+            let ctxt = expr.span.ctxt();
+            let mut applicability = Applicability::MachineApplicable;
+            let trimmed_ok = snippet_with_context(cx, recv.span, ctxt, "", &mut applicability).0;
+            let sugg = format!("let _ = {}", trimmed_ok.trim().trim_end_matches('.'),);
```
[the remark this answers is not in the document]
Comment (ithinuel) [reply clippy_lints/src/misleading_use_of_ok.rs:45]: I'm not 100% sure.
I used matching_result_ok as an example.

IIUC it removes the extra spaces on cases like line 14 of the UI test sample.

File: clippy_lints/src/misleading_use_of_ok.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/misleading_use_of_ok.rs:37]: It should be possible with [`in_external_macro(cx.sess(), stmt.span)`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/lint/fn.in_external_macro.html)

File: clippy_lints/src/misleading_use_of_ok.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/misleading_use_of_ok.rs:45]: 🤔 That's odd, I'm not sure what's going on in match_result_ok. The span of the receiver of a method call should always be
```
    x   .   parse::<i32>()   .   ok   ();
    ^^^^^^^^^^^^^^^^^^^^^^
```
I tried removing the trimming in both this lint and in match_result_ok and in both cases it seems to pass all tests just fine. @flip1995 seeing that you reviewed #5032, which is where that was introduced, do you know why this was necessary?

**Comment by ithinuel** [comment]
I discovered today that `_ = …` is a valid alternative to `let _ = …`.
Personally, I tend to prefer the lighter `_ = …`. However, I can see that the current code base, particularly the UI tests, heavily uses `let _ = …`.
I’m open to either approach, so please let me know if there is a preferred option.

File: clippy_lints/src/misleading_use_of_ok.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (flip1995) [reply clippy_lints/src/misleading_use_of_ok.rs]: No idea anymore. My suspicion is, that back then I/we cared more about well formatted suggestions. Nowadays, we don't really care about that so much, as we assume that everyone uses `rustfmt` anyway and Clippy suggestion formatting is not important.

But that is just a wild guess.

[the remark this answers is not in the document]
Comment (ithinuel) [reply clippy_lints/src/misleading_use_of_ok.rs]: Is there any compatibility requirement with previous version of the compiler?

As far as I understand, `_ = ` is a destructuring assignment introduced in [1.59](https://blog.rust-lang.org/2022/02/24/Rust-1.59.0.html#destructuring-assignments).

[the remark this answers is not in the document]
Comment (flip1995) [reply clippy_lints/src/misleading_use_of_ok.rs]: Do you mean in the Clippy codebase or in the suggestions made by Clippy?

In the former case: no

In the latter case: Yes. If you want to add a suggestion that suggests `_ = ...`, it should probably be gated by an MSRV check for `1.59.0` where that feature was added. There's a lot of examples in the code of how to do that, should it be required here.

**Comment by dswij** [comment]
FME, `let _ = ...` is a much more popular approach. But again, it might be only on the codebases that I've seen, and I don't have real data to back it up. IMO, It's fine to use ` let _ = ...` for now. wdyt? @ithinuel

**Comment by ithinuel** [comment]
I’ll keep it as `let _ = `.

I don’t think having a different recommendation based on the compiler version in used is a great things. The noise caused would outweigh the subjective readability benefits.

File: clippy_lints/src/misleading_use_of_ok.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ithinuel) [reply clippy_lints/src/misleading_use_of_ok.rs]: I meant the suggestion. So yes, let’s keep it as `let _ = …`.
I’ll let you resolve this conversation if it the required changes have been met.

**Comment by y21** [comment]
I agree with @dswij, and FWIW, the rustc lint `unused_must_use` also suggests `let _ = ...`, so having this lint use it as well sounds good?
IMO the lint name is a bit unclear about what exactly it's looking for. What about `unused_result_ok`? That better captures the pattern and makes it easier if in the future we want more lints for other misleading uses of `.ok()`. It's also similar to `match_result_ok`.

File: tests/ui/misleading_use_of_ok.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+fn good_style(x: &str) -> Option<u32> {
+    x.parse::<u32>().ok()
+}
+
+#[rustfmt::skip]
+fn strange_parse(x: &str) {
+    x   .   parse::<i32>()   .   ok   ();
+}
+
```
Comment (y21) [inline tests/ui/misleading_use_of_ok.rs:16]: Would be good to have some test cases involving macros. Something like
```rs
macro_rules! v {
  () => { Ok::<(), ()>(()) }
}

v!().ok();
```
to exercise the case where only the receiver of the `.ok()` call is from a macro (this sometimes trips up lint suggestions, in my experience), but the current implementation should get this right, given that it's using the syntax context of the `ok()` call expression for the snippet

Comment (y21) [reply tests/ui/misleading_use_of_ok.rs:16]: And also one where the whole `stmt.ok();` call is from an external macro, to make sure that checks works correctly. You can take a look at e.g. [this test](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/unneeded_field_pattern.rs) how it does that. The `external!` macro "tricks" clippy into thinking that it's from an external crate

Comment (ithinuel) [reply tests/ui/misleading_use_of_ok.rs]: I did implement tests with internal and `external!` macros.
`is_external_macros` seems to not be fooled by the `external!` trick.
I pushed the code anyway, maybe you’ll see my mistake (I barely understand how those things really work).

Comment (y21) [reply tests/ui/misleading_use_of_ok.rs]: Not entirely sure what's going on there. I'm guessing that it's something to do with the nested `macro_rules!` inside the `external!`. Does it work (i.e. not lint) if you just have something like `external! { Ok::<(), ()>(()).ok(); }`?

Comment (ithinuel) [reply tests/ui/misleading_use_of_ok.rs]: Yes. `external! { Ok::<(), ()>(()).ok(); }` is not reported by this lint.

**Comment by ithinuel** [comment]
@y21 & @flip1995  is there anything blocking the PR from making it through?

File: clippy_lints/src/unused_result_ok.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+use rustc_errors::Applicability;
+use rustc_hir::{ExprKind, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Result::ok();` when the result is not captured.
```
Comment (y21) [inline clippy_lints/src/unused_result_ok.rs:13]: I find the word "captured" a bit ambiguous here. How about something like, "Checks for calls to `Result::ok()` without using the returned `Option`"?

File: clippy_lints/src/unused_result_ok.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    pub UNUSED_RESULT_OK,
+    style,
+    "Use of `.ok()` to silence `Result`'s `#[must_use]` is misleading. Use `let _ =` instead."
+}
+declare_lint_pass!(UnusedResultOk => [UNUSED_RESULT_OK]);
+
+impl LateLintPass<'_> for UnusedResultOk {
+    fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &Stmt<'_>) {
+        if let StmtKind::Semi(expr) = stmt.kind &&
+            let ExprKind::MethodCall(ok_path, recv, [], ..) = expr.kind //check is expr.ok() has type Result<T,E>.ok(, _)
```
Comment (y21) [inline clippy_lints/src/unused_result_ok.rs:39]:
```suggestion
        if let StmtKind::Semi(expr) = stmt.kind
            && let ExprKind::MethodCall(ok_path, recv, [], ..) = expr.kind //check is expr.ok() has type Result<T,E>.ok(, _)
```
🤔 I wonder why rustfmt didn't catch this

**Review by y21 — APPROVED** [review]
Sorry for the long wait. Implementation LGTM, I just have some small documentation nits, then IMO it's merge ready? @dswij do you have something else, since you're assigned?

File: clippy_lints/src/unused_result_ok.rs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            && ok_path.ident.as_str() == "ok"
+            && is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Result)
+            && !in_external_macro(cx.sess(), stmt.span)
+        {
+            let ctxt = expr.span.ctxt();
+            let mut applicability = Applicability::MachineApplicable;
```
Comment (y21) [inline clippy_lints/src/unused_result_ok.rs:45]: Thinking about it, are we sure that ignoring the error is guaranteed to be what the user wants? `MachineApplicable` is a strong assertion that the suggestion is what the user expects. Maybe the user wrote `.ok();` but forgot to actually *do* something with it? I'd be more comfortable with `MaybeIncorrect`, so that tools like rustfix won't blindly use the suggestion, but happy to hear what others think

**Comment by y21** [comment]
We could also just add `#[must_use]` to the ok method in the standard library and that would cause `unused_must_use` to start linting on `.ok();`  but that might be *too* opinionated for a rustc lint.
I could see that leading to people who disagree with it to allow `unused_must_use` entirely, which would be worse because they'd miss definitely-wrong code warnings, like `string.trim();` being a no-op.

So, probably better to have it as a clippy lint instead.

**Comment by Jarcho** [comment]
This definitely looks like a case for `#[must_use]`. An unused result is either an obfuscated drop (`x.ok();` will drop `x`), or a useless method call (`foo().ok()` does nothing `foo()` doesn't). In both cases there's a more idiomatic form to use to accomplish the goal.

**Comment by xFrednet** [comment]
Hey @y21 and @Jarcho, it looks like this PR has been approved, but there was a following discussion. Could you take a look at it?

@rustbot ready

**Comment by y21** [comment]
It probably makes sense to open a discussion for this on zulip. I view rustc's `unused_must_use` lint almost as a correctness lint since it's typically applied to things where it really doesn't make any sense to ignore it and it must be a bug (for example, ignoring futures without doing something with them is a no-op, calling `.trim()` on a `&str` without using it is a no-op), whereas using `.ok()` on a `Result` to suppress the lint is more of a style thing that people might disagree with. To give one concrete example, the `dotenv` crate has `dotenv().ok();` in its example on the readme.


I think adding `#[must_use]` to `ok` would also lead to the suggestion `let _ = result.ok();` when it could have just been `let _ = result;`, so a clippy lint allows for better diagnostics, though `#[must_use]` allows providing a message so that could be easily mitigated.

File: clippy_lints/src/unused_result_ok.rs (hunk 6 above)
Comment (J-ZhengLi) [reply clippy_lints/src/unused_result_ok.rs:45]: agreed, someone might intended to put `some_res.ok()?;` but forgets the question mark, looks like something I would do :D

**Comment by J-ZhengLi** [comment]
Well, it has been another 2 months~ 

so, pinging the most recent people who replies (how you don't mind): @y21 @xFrednet, anyone wanna have the privilege to open a discussion/FCP on zulip ? 😄

At the mean time, @ithinuel, sorry for the long wait, do you mind applying those suggestions made by `y21` on February?

**Comment by xFrednet** [comment]
I can give this another look, if @y21 doesn't get to it during the weekend :)

**Comment by y21** [comment]
Sorry, missed this. Implementation wise this all still looks good to me. There's still the question if we want this in the first place.
FCP: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20.60unused_result_ok.60

**Comment by Alexendoo** [comment]
See also https://github.com/rust-lang/rust/issues/66116

**Comment by xFrednet** [comment]
Hey, this is a ping from triage. @y21 I think the FCP period should be done. Can you summarize the results and move this formward?

@rustbot ready

**Comment by y21** [comment]
Looks like people are generally on board with having this be a clippy lint, based on the FCP thread. The discussion on the issue that Alexendoo linked also shows that it is a pattern that people write and would probably be too strong for rustc to warn on as part of unused_must_use for now, and it's also been tried before in rust-lang/rust#79006 and got rejected.

@flip1995 on Zulip even went as far as wanting this in the restriction category (ie allow-by-default, not `style`) because it's so common. I think I can get behind this, because it looks like people are still figuring out what the ideal solution for this would be and there doesn't appear to be a very clear answer on what is generally considered more idiomatic to the point where we should start pushing it onto people, but ultimately I don't have a strong preference here.
I'll post a poll soon just to get more opinions on that

(There's also the concern that this lint will need to be deprecated when that method gets marked as `#[must_use]` but I don't think it's worth blocking it over that -- we had to do the same just recently with maybe_misused_cfg when check-cfg got stabilized)

**Comment by y21** [comment]
Alright, [the poll] didn't bring in a *lot* more opinions but we did get another vote for restriction, so that sounds like we should go with that.
@ithinuel could you make this last change? I believe that was the last "concern" here and should be ready to go with the category changed. (And also squashing the commits would be nice)

[the poll]: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20.60unused_result_ok.60

**Comment by ithinuel** [comment]
That should do the trick (I also rebased on top of master).

**Comment by y21** [comment]
There's still a bunch of commits that don't seem meaningful outside of this PR (eg "apply review feedback", "remove trim", "rename and add tests", ...). Can you squash those also?

**Comment by y21** [comment]
Also @dswij , I forgot that you're still also assigned to this issue. Did you have anything else?

**Comment by ithinuel** [comment]
@y21, I squashed the commits together and kept the relevant commit title.

**Comment by y21** [comment]
I think `@dswij` is busy, so I'm going to just go ahead and skip that to avoid having to block this for even longer.

Thank you for the effort you put into this @ithinuel ! 

@bors r+
